src/remote.cc

```diff
@@ -682,6 +688,102 @@ static void gr_slideshow_delay(const gchar *text, GIOChannel *, gpointer)
 	options->slideshow.delay = static_cast<gint>(n * 10.0 + 0.01);
 }
 
+static void gr_duplicates_threshold(const gchar *text, GIOChannel *, gpointer)
+{
+	gint thresh;
+	gint n;
+	gint res;
+
+	res = sscanf(text, "%d", &thresh);
+	if (res == 1)
+		{
+		n = thresh;
+		if (n < 0 || n > 100)
+			{
+			printf_term(TRUE, "Image similarity threshold out of range (%d to %d)\n", 0, 100);
+			return;
+			}
+		}
+	else
+		{
+		n = 99;
+		}
+
+	options->duplicates_similarity_threshold = static_cast<guint>(n);
+	DEBUG_0("threshold set to %d", options->duplicates_similarity_threshold);
+}
+
+static void gr_duplicates_program(const gchar *text, GIOChannel *, gpointer)
+{
+	g_strdup(options->duplicates_program);
+	options->duplicates_program = g_strdup(text);
+	DEBUG_0("duplicates program set to \"%s\"", options->duplicates_program);
+}
+
+static void gr_process_duplicates(const gchar *, GIOChannel *, gpointer data)
+{
+	auto remote_data = static_cast<RemoteData *>(data);
+    std::map<std::string, std::unique_ptr<pic>> pics;
+	GList *work = remote_data->file_list;
+	while (work)
+	{
+		FileData *fd = static_cast<FileData *>(work->data);
+        std::string name(fd->path);
+        pics[name] = std::unique_ptr<pic>(new pic(fd->path));
+		work = work->next;
+	}
+	DEBUG_1("processing %d files in set", pics.size());
+
+	// Compute similarity score for every pair, build equivalence sets.
+    for (auto a = pics.begin(); a != pics.end(); ++a) {
+        auto b = a;
+        b++;
+        for (; b != pics.end(); ++b) {
+            double similarity = a->second->compare(*b->second);
+            DEBUG_1("%s vs %s: %f", a->second->name.c_str(), b->second->name.c_str(), similarity);
+            if (similarity < options->duplicates_similarity_threshold)
+                continue;
+            a->second->equivalent.insert(b->second->equivalent.begin(), b->second->equivalent.end());
+            for (auto const &f: a->second->equivalent) {
+                pics[f]->equivalent.insert(a->second->equivalent.begin(), a->second->equivalent.end());
+            }
+        }
+    }
+
+    std::set<std::string> printed;
+    for (auto const &f: pics) {
+        if (f.second->equivalent.size() < 2)
+            // skip this pic if not similar to any other one but itself
+            continue;
+        if (printed.find(f.second->name) != printed.end())
+            // skip this pic if it was already printed (when processing a similar image)
+            continue;
+        std::vector<const char *> cmd;
+		cmd.push_back(options->duplicates_program);
+        for (auto const &e: f.second->equivalent) {
+            cmd.push_back(e.c_str());
+            printed.insert(e);
+        }
+        cmd.push_back(NULL);
+        pid_t pid = fork();
+        if (pid == -1) {
+            perror("fork");
+            exit(1);
```

Comment:
`exit()` is not called directly from anywhere else in this file, and geeqie does not generally use `perror`.  The best course of action here is likely to log an error and return.  See an example here: \
https://github.com/BestImageViewer/geeqie/blob/8cdab9999d427756f8e0ab3616f854938796ff5c/src/remote.cc#L541

With that said, it's probably a lot safer to use an API like https://docs.gtk.org/gio/ctor.Subprocess.new.html instead of `fork`ing directly.  That said, @qarkai is our resident expert on glib APIs, so I would defer to him.

Reply:
I haven't used GSubprocess much. geeqie uses `g_spawn_async_with_pipes` for opening external editor, see `editor_command_one` in `src/editors.cc:1053`.

Reply:
Geeqie uses `g_spawn_async_with_pipes()` for external editor. See `editor_command_one()` in `src/editors.cc:1053`.